src/merge.rs

```diff
@@ -186,13 +187,17 @@ pub fn merge(
             let body = Term::ContractWithDefault(ty_closure, lbl, body).into();
             Ok(Closure { body, env })
         }
-        (Term::ContractWithDefault(ty1, lbl1, t1), Term::ContractWithDefault(ty2, _lbl2, t2)) => {
-            //FIXME: The choice of lbl1 is totally arbitrary, to please the compiler. Ideally
-            //labels should also be mergeable, but the current PR is already getting too big, and
-            //this is left for future work
+        // Composing two contracts creates a new one which applies these contracts successively.
+        // This composite contract requires its own label, but its components already had their
+        // own, hence they will ignore this label. We make the arbitrary choice of passing the
+        // first label `lbl1` in each such situation: because it will be ignored, it doesn't
+        // matter.
+        (Term::ContractWithDefault(ty1, lbl1, t1), Term::ContractWithDefault(ty2, lbl2, t2)) => {
             let Closure { body, mut env } = mk_merge_closure(t1, env1.clone(), t2, env2.clone());
             let body = Term::ContractWithDefault(
-                merge_types_closure(&mut env, ty1, env1, ty2, env2),
+                merge_types_closure(&mut env, ty1, lbl1.clone(), env1, ty2, lbl2, env2),
```

Comment:
If you are going to not use the other copy of `lbl1`, maybe it's better to avoid a potentially non-trivial call to `clone` and instead, make a truly dummy label with no content.